src/streaming-services/common/ScrobbleEvents.ts

```diff
@@ -1,59 +1,131 @@
 import { EventDispatcher } from '../../common/Events';
+import { ScrobbleParser } from './ScrobbleParser';
+
+export interface ScrobbleEventsOptions {
+	/**
+	 * How frequently to check for changes, in seconds.
+	 *
+	 * *Default:* `0.5`
+	 */
+	checkFrequency: number;
+}
 
 export abstract class ScrobbleEvents {
-	changeListenerId: number | null;
-	isPaused: boolean;
-	isPlaying: boolean;
+	protected parser: ScrobbleParser;
+	readonly options: Readonly<ScrobbleEventsOptions>;
+	protected changeListenerId: number | null = null;
+	protected url = '';
+	protected playbackStarted = false;
+	protected isPaused = true;
+	protected progress = 0.0;
 
-	constructor() {
-		this.changeListenerId = null;
-		this.isPaused = false;
-		this.isPlaying = false;
+	constructor(parser: ScrobbleParser, options: Partial<ScrobbleEventsOptions> = {}) {
+		this.parser = parser;
+		this.options = Object.freeze({
+			...this.getDefaultOptions(),
+			...options,
+		});
 	}
 
-	startListeners = () => {
-		this.addChangeListener();
-	};
-
-	stopListeners = () => {
-		this.stopChangeListener();
-	};
+	protected getDefaultOptions(): ScrobbleEventsOptions {
+		return {
+			checkFrequency: 0.5,
+		};
+	}
 
-	getLocation = (): string => {
+	protected getLocation(): string {
 		return window.location.href;
-	};
+	}
 
-	addChangeListener = () => {
+	startListeners() {
 		void this.checkForChanges();
-	};
+		if (this.parser.onClick) {
+			document.addEventListener('click', this.parser.onClick, true);
+		}
+	}
 
-	stopChangeListener = () => {
+	stopListeners() {
 		if (this.changeListenerId !== null) {
 			window.clearTimeout(this.changeListenerId);
+			this.changeListenerId = null;
+		}
+	}
+
+	protected async checkForChanges(): Promise<void> {
+		const newUrl = this.getLocation();
+		if (this.url !== newUrl) {
+			await this.onUrlChange(this.url, newUrl);
+			this.url = newUrl;
+		}
+
+		const playback = await this.parser.parsePlayback();
+		if (playback) {
+			const newProgress = playback.progress;
+			if (this.progress !== newProgress) {
+				await this.updateProgress(newProgress);
+				this.progress = newProgress;
+			}
+
+			try {
+				if (playback.isPaused && !this.isPaused) {
+					await this.pause();
+				} else if (!playback.isPaused && this.isPaused) {
```

Comment:
It should not start a scrobble if item is not present. If the Item is retrieved later it will not (re)start the scrobble, and the only way to make it scrobble is to manually stop/start the player. Not sure if you want to prevent the playback from returning a value if item is not found, or if you want to include item in this file.

Reply:
Oh yeah, `parsePlayback` should return `null` if there's no item. I realize that I only checked for that before parsing the item, but not after to make sure that the item was actually parsed.